File: coding/Leetcode/Dynamic_Programming/task_300.cpp
// https://leetcode.com/problems/longest-increasing-subsequence/
// https://leetcode.com/problems/longest-increasing-subsequence/
// Created by ruslan515 on 19.12.22.
//

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int ans = 0;
        int n = nums.size();
        vector<int> dp(n + 1, 0);
        dp[0] = 0;

        int global_max = dp[0];
        for (int i = 1; i < n + 1; ++i) {
            int max_dp = dp[0];
            int t1 = nums[i - 1];
            for (int j = 0; j < i; ++j) {
                if (nums[j] < t1) {
                    max_dp = max(max_dp, dp[j + 1]);
                }
            }
            dp[i] = max_dp + 1;
            global_max = max(global_max, dp[i]);
        }
        ans = global_max;
        return ans;
    }
};

void read_data(vector<int>& nums) {
    int temp;
    while (cin >> temp)
        nums.push_back(temp);
    return;
}

int main() {
    vector<int> nums;
    read_data(nums);
    Solution solve;
    cout << solve.lengthOfLIS(nums);

    return 0;
}

File: coding/Leetcode/Dynamic_Programming/task_300.py
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0 for _ in range(n)]
        dp[0] = 1
        answer = dp[0]

        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j])
            dp[i] += 1
            answer = max(answer, dp[i])

        return answer


def main():
    solve = Solution()

    nums = [0]
    answer = 1
    assert answer == solve.lengthOfLIS(nums)

    nums = [10, 9, 2, 5, 3, 7, 101, 18]
    answer = 4
    assert answer == solve.lengthOfLIS(nums)

    nums = [0, 1, 0, 3, 2, 3]
    answer = 4
    assert answer == solve.lengthOfLIS(nums)

    nums = [7, 7, 7, 7, 7, 7, 7]
    answer = 1
    assert answer == solve.lengthOfLIS(nums)


if __name__ == "__main__":
    main()
